spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java

```diff
@@ -173,7 +173,7 @@ protected Table newStaticTable(TableMetadata metadata, FileIO io) {
 
   protected Dataset<Row> buildOtherMetadataFileDF(TableOperations ops) {
     List<String> otherMetadataFiles = getOtherMetadataFilePaths(ops);
-    return spark.createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path");
+    return spark.createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path").distinct();
```

Comment:
Instead of doing a shuffle here, I think we should refine the approach we use to build the list of JSON files. I think what happens now is that we will take previous 100 version files in every version file and add them to the list even though each new version file has only one different entry. Will tables with 2000 snapshots and 100 previous metadata files produce a list with 200000 elements?

Reply:
@aokolnychyi This current code does not do a recursive listing of all the version metadata json files. 
I have made changes to https://github.com/apache/iceberg/pull/2415 to do a recursive listing and also deduped the files as part of the change.